docs/concepts/architecture.md

```diff
@@ -113,3 +113,43 @@ Nuclio provides the ability to run different versions of the same function simul
 
 The event-source mapping can either specify the exact function version, or utilize a defined alias and thus eliminate the need to change the mapping when a newer version of the function is published.
 
+
+## Processors' internal abstractions
+
+Here is an illustration of the abstraction levels that every event goes through while being processed:
+![event-processing-flow](../../docs/assets/images/internal-processor-abstractions.png)
+
+Thus, every event goes through two EventProcessor levels: the first at the trigger level and the second at the runtime level.
+The EventProcessor exists to generalize the allocation logic and make it abstract.
+
+The event processor allocation logic varies based on the expected behavior. We support two modes:
+Synchronous (default) – Events within a single worker are processed sequentially (FIFO).
+Asynchronous – Events within a single worker are processed concurrently.
+
+Depending on the required behavior, we use different event processor allocators at different levels.
+
+For instance, to support synchronous processing, we allocate a worker (event processor at the trigger level).
+While the allocated event processor is waiting for a response, it cannot be allocated again. To enforce this, we use a sync pool allocator,
+which ensures that each event processor is assigned to only one event at a time.
+
+For asynchronous processing, a worker can be allocated for multiple events simultaneously.
+In this case, we use a singleton async allocator, which allows multiple events to share the same event processor without blocking further allocations.
+However, as the event moves further through the processing flow, we still need to allocate an event processor at the runtime level. While an event is being processed, we must ensure that no other events are processed through the same event processor to maintain correct behaviour.
```

Comment:
This sentence isn't clear (to me). Do you mean:
However, as the event moves through the processing flow, we still need to allocate an event processor at the runtime level, it is allocated an event processor that is not used by any other events.

Reply:
@jillnogold not really, I meant that the event processor  Nuclio allocated on runtime level, may be used to process only one event at the same time. So nuclio ensures that no other events are processed through the same event processor to maintain correct behaviour.

Reply:
not sure how to make it better. Maybe smth like:

```
While an event is being processed by allocated event processor on runtime level, Nuclio ensures that no other events are processed through the same event processor to maintain correct behaviour.

```